api/cpp/include/slint.h

```diff
@@ -1049,6 +1049,182 @@ class MapModel : public Model<MappedModelData>
+            ++it;
+        }
+
+        for (int removed_row : removed_rows) {
+            target_model.row_removed(removed_row, 1);
+        }
+    }
+    void reset() override
+    {
+        sorted_rows_dirty = true;
+        target_model.reset();
+    }
+
+    void ensure_sorted()
+    {
+        if (!sorted_rows_dirty) {
+            return;
+        }
+
+        sorted_rows.resize(source_model->row_count());
+        for (size_t i = 0; i < sorted_rows.size(); ++i)
+            sorted_rows[i] = i;
+
+        std::sort(sorted_rows.begin(), sorted_rows.end(), [this](int lhs_index, int rhs_index) {
+            auto lhs_elem = source_model->row_data(lhs_index);
+            auto rhs_elem = source_model->row_data(rhs_index);
+            return sort_fn(*lhs_elem, *rhs_elem);
+        });
+
+        sorted_rows_dirty = false;
+    }
+
+    std::shared_ptr<slint::Model<ModelData>> source_model;
+    std::function<bool(const ModelData &, const ModelData &)> sort_fn;
+    slint::SortModel<ModelData> &target_model;
+    std::vector<int> sorted_rows;
+    bool sorted_rows_dirty = true;
+};
+}
+
+/// The SortModel acts as an adapter model for a given source model by sorting all rows
+/// with by order provided by the given sorting function. The sorting function is called for
+/// pairs of elements of the source model.
+template<typename ModelData>
+class SortModel : public Model<ModelData>
+{
+    friend struct private_api::SortModelInner<ModelData>;
+
+public:
+    /// Constructs a new SortModel that provides a sorted view on the \a source_model by applying
+    /// the order given by the specified \a sort_fn.
+    SortModel(std::shared_ptr<Model<ModelData>> source_model,
+              std::function<bool(const ModelData &, const ModelData &)> sort_fn)
+        : inner(std::make_shared<private_api::SortModelInner<ModelData>>(std::move(source_model),
+                                                                         std::move(sort_fn), *this))
+    {
+        inner->source_model->attach_peer(inner);
+    }
+
+    int row_count() const override { return inner->source_model->row_count(); }
+
+    std::optional<ModelData> row_data(int i) const override
+    {
+        inner->ensure_sorted();
+        return inner->source_model->row_data(inner->sorted_rows[i]);
+    }
+
+    /// Re-applies the model's sort function on each row of the source model. Use this if state
+    /// external to the sort function has changed.
+    void sort() { inner->reset(); }
+
+    /// Given the \a sorted_row_index, this function returns the corresponding row index in the
+    /// source model.
+    int unsorted_row(int sorted_row_index) const
+    {
+        inner->ensure_sorted();
+        return inner->sorted_rows[sorted_row_index];
+    }
+
+    /// Returns the source model of this filter model.
+    std::shared_ptr<Model<ModelData>> source_model() const { return inner->source_model; }
+
+private:
+    std::shared_ptr<private_api::SortModelInner<ModelData>> inner;
```

Comment:
Also valid for other adapter:  Why is it a shared_ptr?  Should it not be a unique_ptr, or even just a plain member?
I also wonder if the two class couldn't be merged in one by using private inheritance from ModelChangeListener

Reply:
The reason for it being a shared_ptr is because it also acts as `ModelPeer`, which is a `std::weak_ptr<ModelChangeListener>`.

If we merge it, then I think we need `std::enable_shared_from_this` and basically replace the constructor for the models with a factor function that returns a `std::shared_ptr`.

I can draft a diff after this change and then we can see how it looks like and compare.